relations/src/shielder/tangle.rs

```diff
@@ -1,67 +1,85 @@
-//!  1.3 We go bottom-to-top and in every intermediate node we:
+//! Tangling is a function that takes in a sequence of field elements (either raw `CircuitField`s
+//! ([tangle]) or as `FpVar`s ([tangle_in_circuit])) and mixes it into a single field element.
+//!
+//! It operates in three steps:
+//!  1.1 We repeat the sequence until the result has [EXPAND_TO] elements. If the input sequence is
+//!      longer than [EXPAND_TO], it will be trimmed.
+//!  2.1 For every chunk of length `BASE_LENGTH` we compute _spiced_ suffix sums of inverses.
+//!     Basically, apart from just summing element inverses, we multiply intermediate results by an
+//!     index-dependent factor.
+//!  2.2 We build a binary tree over these chunks.
+//!  2.3 We go bottom-to-top and in every intermediate node we:
 //!      1.3.1 swap the halves
 //!      1.3.2 compute prefix products
-//!  2.1 Given a new mangled sequence of `n` elements we squash it `SQUASH_FACTOR` times, i.e. we
-//!      take chunks of length `SQUASH_FACTOR` and reduce them to a single byte by xoring.
+//!  3.1 A new mangled sequence of `n` elements is reduced by summing.
+//!
+//! In some places, where an element turns out to be zero, we replace it by an index-dependent
+//! constant.
 //!
 //! Note, it is **not** hiding like any hashing function.
 //!
-//! This module exposes two implementations of tangling: `tangle` and `tangle_in_field`. They are
+//! This module exposes two implementations of tangling: [tangle] and [tangle_in_circuit]. They are
 //! semantically equivalent, but they just operate on different element types.
 //!
 //! All the index intervals used here are closed-open, i.e. they are in form `[a, b)`, which means
 //! that we consider indices `a`, `a+1`, ..., `b-1`. We also use 0-based indexing.
 
-use ark_r1cs_std::R1CSVar;
+use std::ops::Add;
+
+use ark_ff::{Field, Zero};
+use ark_r1cs_std::fields::FieldVar;
 use ark_relations::r1cs::SynthesisError;
-use ark_std::vec::Vec;
 
-use super::types::ByteVar;
+use crate::{environment::FpVar, CircuitField};
 
 /// Bottom-level chunk length.
 const BASE_LENGTH: usize = 4;
+/// Tangling operates on fixed-sized arrays with 128 elements.
+const EXPAND_TO: usize = 128;
 
-/// Tangle elements of `bytes`.
+/// Entangle `input` into single `FpVar`.
 ///
 /// For circuit use only.
-pub(super) fn tangle_in_field<const SQUASH_FACTOR: usize>(
-    mut bytes: Vec<ByteVar>,
-) -> Result<Vec<ByteVar>, SynthesisError> {
-    let number_of_bytes = bytes.len();
-    _tangle_in_field(&mut bytes, 0, number_of_bytes)?;
-    Ok(bytes
-        .chunks(SQUASH_FACTOR)
-        .map(|chunk| {
-            chunk
-                .iter()
-                .cloned()
-                .reduce(|x, y| x.xor(&y).unwrap())
-                .unwrap()
-        })
-        .collect())
+pub(super) fn tangle_in_circuit(input: &[FpVar]) -> Result<FpVar, SynthesisError> {
+    let mut input_expanded = input
+        .iter()
+        .cycle()
+        .take(EXPAND_TO)
```

Comment:
So this method works only on 128 elements? Anything above them will get thrown away?

Reply:
yes, but if you are trying to make some tangling on >128 elements in your circuit, than it might be infeasible and should definitely be changed